Follow-up to #2

bbguimaraes opened the pull request:
Import templates from origin-metrics, persistent volumes, cassandra cluster.

bbguimaraes committed:
Minor adjustments.

- heapster_allowed_users configuration
- remove `loop_var`s

bbguimaraes committed:
Import templates from origin-metrics.

bbguimaraes committed:
Add option to deploy with persistent volumes.

bbguimaraes committed:
Cassandra cluster.

jcantrill commented on roles/openshift_metrics/files/templates/cassandra_serviceaccount.yaml:
```diff
@@ -0,0 +1,8 @@
+apiVersion: v1
```
so in logging... template files are found in <role>/templates  instead of <role>/files/templates.  this way you can specify src=template_name without having to define explicitly where it is located.

bbguimaraes replied:
There are two kinds of templates in this role, it's a bit unfortunate that they are both called "templates".  These files in `roles/openshift_metrics/files/templates` are static openshift/kubernetes templates that are directly `oc create`d.

The others are (ansible) templates in `roles/openshift_metrics/templates` that will be processed by the `template` tasks in the playbooks to expand (ansible) variables.

jcantrill replied:
I see the distinction here.  You may have answered this in a following comment, but maybe we should consider a naming convention instead of dropping them in a sub-folder if we wish to keep a static openshift template. Something like:  files/cassandra_serviceaccount_template.yaml

jcantrill commented on roles/openshift_metrics/tasks/generate_certificates.yaml:
```diff
@@ -126,7 +124,7 @@
 - name: read files for the hawkular-metrics secret
   command: >
-    base64 --wrap 0 "{{ mktemp.stdout }}/certs/{{ file }}"
+    base64 --wrap 0 "{{ mktemp.stdout }}/certs/{{ item }}"
```
@ewolintez has recommanded we migrate to using 'loop_variable' to override the use of 'item' in order to defend against item being populated somewhere else.

bbguimaraes replied:
The only problem I found was when a loop was used in an `include`, which causes the override problem.  What I did was change the `include`'s `loop_variable`.  And it looked like ansible doesn't show the value of the loop variable on the output unless it is called `item` (but I didn't test that extensively).

jcantrill commented on roles/openshift_metrics/tasks/generate_serviceaccounts.yaml:
```diff
@@ -1,17 +1,15 @@
       metrics-infra: support
     secrets:
-    - hawkular-{{component}}-secrets
+    - hawkular-{{item}}-secrets
```
see comment recarding loop_control

jcantrill commented on roles/openshift_metrics/tasks/install_hawkular.yaml:
```diff
@@ -1,73 +1,35 @@
-#    | oc create -f -
-#  with_sequence: start=1 end={{ cassandra_nodes }}
-#  when: {{ cassandra_nodes > 1 }}
+  with_sequence: start=1 end={{ hawkular_cassandra_nodes }}
```
you could use 'count' here instead of start and end

bbguimaraes replied:
Right, this is a left-over from before when there was a separate step for the master and this was `start=2`.  I will correct it.

jcantrill commented on roles/openshift_metrics/tasks/install_metrics.yaml:
```diff
@@ -7,9 +7,10 @@
-- name: Create temp directory for all our templates
-  file: path={{mktemp.stdout}}/templates state=directory mode=0755
-  changed_when: False
+- name: copy `templates` directory
```
In logging...i think this task is equivalent to 'generate all my objects that i will create later'  Are we certain we need this?

jcantrill commented on roles/openshift_metrics/vars/main.yaml:
```diff
@@ -18,3 +18,8 @@ hawkular_persistence: "{% if openshift.hosted.metrics.storage.kind != None %}tru
 
 metrics_upgrade: openshift.hosted.metrics.upgrade | default(False)
+
+hawkular_cassandra_storage_types:
```
Should this go in defaults/main.yaml

bbguimaraes replied:
This is not an input to the playbook (that is `hawkular_cassandra_storage_type`, without the 's').  These are just the valid values for that variable, it is used for validation (see `roles/openshift_metrics/tasks/main.yaml`).

That said, I don't know if it justifies having it in `vars`.

ewolinetz replied:
I think this makes sense to keep in `vars` for this reason

ewolinetz commented on roles/openshift_metrics/files/templates/hawkular_cassandra_nodes_service.yaml:
```diff
@@ -0,0 +1,20 @@
+apiVersion: v1
```
do we want to use a service template similar to what we use for logging with these services?

ewolinetz commented on roles/openshift_metrics/files/templates/cassandra_serviceaccount.yaml:
```diff
@@ -0,0 +1,8 @@
+apiVersion: v1
```
same question asked regarding services and templates for serviceaccounts

ewolinetz commented:
other than questions regarding using more generic templates LGTM

bbguimaraes commented:
Regarding the `files/templates` comments, this is something I'm not sure about.  My idea was that doing it this way you can see the objects that will be created as yaml files, instead of having to look at the ansible playbooks.

ewolinetz commented:
My thought with using `templates/` over `files/templates` was to reduce the number of files that we need to provide and maintain. Especially since most of the sections are the same, it makes sense to leverage jinja templates to build it out for us.

I would definitely agree that this way is easier to see what is going on. The hope was that the variable names we use for the templates gives us an idea what is actually going into the template (e.g. ports, labels, selectors)

bbguimaraes commented:
Looking at the actual yaml files now, I don't feel it's really necessary to have them anymore, they are so simple.  And one can quickly grep for the template name and find the places where it's used.

I'll change them to ansible templates.

P.S.: on Monday ;)

bbguimaraes committed:
Service creation using a template.

bbguimaraes committed:
Service account creation using a template.

bbguimaraes committed:
Updates for origin-metrics master.

bbguimaraes commented:
I'm glad I didn't say "next Monday" =)

As mentioned, removed the static service and service account yaml files in
favor of using the ansible templates in `roles/openshift_logging`.  I used
symbolic links, from what I remember that is the way we handle these cases.

I had to make a slight change to the `service.j2` template to support headless
services, but it should not affect any existing uses of that template.

The `files/templates` directory no longer exists (and now I realize the files
there were not actually templates, just objects).

Lastly, I had to add some steps (separated on the last commit) to catch up with
the latest changes in origin-metrics (these were what took most of the time, to
be honest).

After review, I believe this is ready to be merged.

jcantrill commented on roles/openshift_metrics/tasks/cleanup.yaml:
```diff
@@ -17,3 +17,14 @@
+  command: >
+    {{ openshift.common.client_binary }} get rolebindings
+    -o jsonpath='{.items[?(@.roleRef.name == "view")].userNames}'
+  register: ret
```
Scope of variables is still a little bit of a mystery to me so, does this need to be something more descriptive and unique?

ewolinetz replied:
this looks odd because its the jsonpath and there isn't a really easy way to substitute this other than using --template (but that'll get a little ugly with needing to escape the '{{' '}}'). 
This looks over all the records returned from 'oc get rolebindings' and for any that have a RoleRef.name == "view" it will return the userNames.

jcantrill replied:
I'm actually concerned about calling it 'ret'.  Does it need to be expanded to something like 'metrics_rb_usernames' or something?

jcantrill commented on roles/openshift_metrics/tasks/generate_certificates.yaml:
```diff
@@ -119,14 +117,17 @@
 - name: generate password for htpasswd file for hawkular metrics
   shell: cat /dev/urandom | tr -dc _A-Z-a-z-0-9 | head -c15
   register: hawkular_metrics_password
+- name: generate password for hawkular metrics jgroups
```
Possibly a later refactoring to keep in mind that may be of use: 
https://github.com/ewolinetz/openshift-ansible/blob/logging_deployer_tasks/roles/openshift_logging/filter_plugins/generators.py#L6  

I added this for generating the variable bit of our DC names.  I'm not sure if there is any advantage or desire to try to use modules and filters over shells and commands

jcantrill commented on roles/openshift_metrics/tasks/generate_serviceaccounts.yaml:
```diff
@@ -1,20 +1,20 @@
 ---
 - name: Generating serviceaccounts
-  template: src={{template_dir}}/serviceaccount.j2 dest={{mktemp.stdout}}/templates/metrics-{{component}}-sa.yaml
+  template: src=serviceaccount.j2 dest={{mktemp.stdout}}/templates/metrics-{{item.name}}-sa.yaml
```
use 'obj_name' since you have it as a var

jcantrill commented on roles/openshift_metrics/tasks/install_hawkular.yaml:
```diff
@@ -1,73 +1,46 @@
+  command: >
+    {{ openshift.common.client_binary }} get rolebindings
+    -o jsonpath='{.items[?(@.roleRef.name == "view")].userNames}'
+  register: ret
```
I question this name choice because of scope considerations

jcantrill commented on roles/openshift_metrics/tasks/install_hawkular.yaml:
```diff
@@ -1,73 +1,46 @@
-  - support
+  - hawkular_metrics_rc
+  - hawkular_cassandra_rc
+- name: test view permission for the hawkular service account
```
In logging we added a logging specific fact gathering step that runs first.  I wonder if that would be of use in this situation.

jcantrill commented on roles/openshift_metrics/tasks/install_hawkular.yaml:
```diff
@@ -1,73 +1,46 @@
+  command: >
+    '{{ openshift.common.admin_binary }}' policy
+    -n '{{ project }}' add-role-to-user view hawkular
+  when: "'system:serviceaccount:{{ project }}:hawkular' not in ret.stdout"
```
we will probably need to change 'project' to be something like 'metrics_project' or 'hawkular_project'

jcantrill commented on roles/openshift_metrics/tasks/install_hawkular.yaml:
```diff
@@ -1,73 +1,46 @@
   command: >
     {{ openshift.common.client_binary }} -n '{{ project }}'
-    create -f '{{ mktemp.stdout }}/templates/{{ file }}'
+    create -f '{{ mktemp.stdout }}/templates/{{ item }}'
```
I think we are using apply in logging because it will not fail if the object already exists.  Also, do we need to explicitly define the list of object to create versus https://github.com/ewolinetz/openshift-ansible/blob/logging_deployer_tasks/roles/openshift_logging/tasks/main.yaml#L38 which uses a file glob

jcantrill commented on roles/openshift_metrics/tasks/install_heapster.yaml:
```diff
@@ -16,11 +16,9 @@
   command: >
     {{ openshift.common.client_binary }} -n '{{ project }}'
-    create -f '{{ mktemp.stdout }}/templates/{{ file }}'
+    create -f '{{ mktemp.stdout }}/templates/{{ item }}'
```
Can we create the object in one step at the end or is there a need to control cassandra before heapster?

jcantrill commented on roles/openshift_metrics/templates/hawkular_cassandra_rc.j2:
```diff
@@ -0,0 +1,128 @@
+apiVersion: v1
+kind: Template
```
Any reason this one remains a template?  Is it because its easier to process the PVC with the RC

bbguimaraes committed:
Better names for register variables.

Scope is the entire playbook, not the individual file as previously thought.

bbguimaraes committed:
Rename `project` to `metrics_project`.

bbguimaraes committed:
Minor fixes.

- remove no longer used template
- reuse variable value
- use a relative path on mktemp to respect $TMPDIR
- headless services change in kubernetes (`portalIP` renamed to `clusterIP`)

ewolinetz commented on roles/openshift_metrics/tasks/install_hawkular.yaml:
```diff
@@ -1,73 +1,46 @@
+  changed_when: false
+- name: add the view role to the hawkular service account
+  command: >
+    '{{ openshift.common.admin_binary }}' policy
```
you don't need to surround this with quotes here, same with -n {{project}}

ewolinetz commented on roles/openshift_metrics/templates/hawkular_metrics_rc.j2:
```diff
@@ -12,48 +12,64 @@ spec:
-        - "-Dorg.apache.catalina.connector.CoyoteAdapter.ALLOW_BACKSLASH=true"
-        - "-Dhawkular-metrics.openshift.auth-methods=openshift-oauth,htpasswd"
-        - "-Dhawkular-metrics.openshift.htpasswd-file=/secrets/hawkular-metrics.htpasswd.file"
+        - "-Dhawkular.metrics.cassandra.nodes=hawkular-cassandra"
```
would it make sense to move these all into a vars/main.yaml entry?

jcantrill replied:
I don't think necessarily unless there is something variable about them.  Previously they were hardcoded into an openshift template so I think its ok to do the same here.

ewolinetz commented on roles/openshift_metrics/tasks/generate_certificates.yaml:
```diff
@@ -37,13 +37,11 @@
   register: sa_secret
 - name: read files for the heapster secret
-  command: base64 --wrap 0 "{{ mktemp.stdout }}/certs/heapster.{{ file }}"
+  command: base64 --wrap 0 "{{ mktemp.stdout }}/certs/heapster.{{ item }}"
```
one change we did recently with logging is to use take advantage of the generated_certs directory that is created when installing OSE. Not sure if you'd want to take the same approach here (we keep our api objects in mktemp.stdout and then only have our certs in the generated_certs dir so we can reuse our signing cert, etc)

bbguimaraes committed:
Changes for hawkular-metrics 0.21.4.

bbguimaraes committed:
PersistentVolumeClaim template.

bbguimaraes committed:
Extract hawkular-cassandra rc from template.

bbguimaraes committed:
Create objects at the end, using a wildcard.

bbguimaraes committed:
Route template.

bbguimaraes committed:
Role binding template.

bbguimaraes commented:
@ewolinetz, @jcantrill: addressed comments, PTAL again.

ewolinetz commented on roles/openshift_metrics/defaults/main.yaml:
```diff
@@ -4,8 +4,14 @@ image_version: latest
+hawkular_cassandra_pv_size: 10Gi
 
 heapster_standalone: False
+heapster_allowed_users: system:master-proxy
```
this should be a comma delimited string correct or is only one user allowed for Heapster? 
I haven't checked the README yet, but we should make sure that's explained otherwise someone may use an array (not sure if that'd still work in secret generation)

bbguimaraes replied:
Agreed.  The README was still the original one, so I went ahead and updated it
now that we have a semi-stable version.  I'm not sure about some of the things
there (license?  authors?), so I didn't touch them.

We have a lot of documentation on these parameters (and other deployer-related
topics) in [origin-metrics](https://github.com/openshift/origin-metrics), and logging is probably similar.
Are we keeping them there?  Or maybe documenting them here or in
openshift-docs?

ewolinetz replied:
My thought is that if its pertinent to the role or running it in the playbook, it should be described here. I'm sure we'll have some overlap for a while between here and origin-metrics and as the use of the deployer drops off and this role is used more for installing we'll start to see things removed from origin-metrics.

I'm not sure the policy on describing roles in the docs, but that would make sense as well. At a minimum I'd feel they would be described in documentation for the enterprise (and I suppose origin too) community, in some capacity.

ewolinetz commented on roles/openshift_metrics/tasks/main.yaml:
```diff
@@ -1,20 +1,24 @@
+- name: create objects
+  command: >
+    {{ openshift.common.client_binary }} -n '{{ metrics_project }}'
+    create -f {{ item }}
```
i would recommend apply instead, create can fail if an object of the same name already exists; apply will try to resolve collisions and update if necessary

bbguimaraes replied:
Changed.  There is still a lot of extra work being done (e.g. secret generation
that's not used) if you rerun the playbook, but at least it doesn't give you an
error and the components keep working.  This is likely what I'll tackle next.

ewolinetz reviewed (commented):
one question/documentation todo and suggestion, otherwise LGTM

bbguimaraes committed:
Update README.

bbguimaraes committed:
Use `oc apply` instead of `oc create`.

ewolinetz commented on roles/openshift_metrics/README.md:
```diff
@@ -5,37 +5,65 @@ OpenShift Metrics Installation
 --------------
 
-From this role:
+For default values, see `defaults/main.yaml`.
```
I think we should describe the defaults here... I realize it's a duplicate, that or make this a link to defaults/main.yaml

jcantrill commented on roles/openshift_logging/templates/rolebinding.j2:
```diff
@@ -1,12 +1,21 @@
 apiVersion: v1
-kind: RoleBinding
+kind: {% if cluster is defined and cluster %}Cluster{% endif %}RoleBinding
```
I should rebase the branch on the latest logging where we have created a 'clusterrolebinding' template.  Sorry for being lax

jcantrill commented on roles/openshift_metrics/tasks/generate_certificates.yaml:
```diff
@@ -118,13 +118,20 @@
   register: hawkular_metrics_password
 - name: generate password for hawkular metrics jgroups
-  shell: cat /dev/urandom | tr -dc _A-Z-a-z-0-9 | head -c15 | base64
+  shell: cat /dev/urandom | tr -dc _A-Z-a-z-0-9 | head -c15
```
We added a filter to logging to do something similiar that would not require you to 'shell'.  It might be worth as a take-away to review and see if it is worth replacing

ewolinetz replied:
this is to generate a random 15 characters, we wouldn't want to actually cat /dev/urandom with that filter...

jcantrill commented on roles/openshift_metrics/tasks/install_hawkular.yaml:
```diff
@@ -1,50 +1,57 @@
-    --dest-ca-cert='{{ mktemp.stdout }}/certs/ca.crt'
+    src: hawkular_metrics_rc.j2
+    dest: "{{ mktemp.stdout }}/templates/hawkular_metrics_rc.yaml"
+- name: generate hawkular-cassandra replication controllers
```
@ewolinetz  We should sync on the approach of doing this and what is done in logging.  We should be consistent and use one or the other.  Logging we create a pool of names that are reused or updated based on 'scaling' the component.  I assume we have similar

ewolinetz replied:
i agree with that, but I didn't think they have a random name as part of any of their components as we do for ES

bbguimaraes replied:
In metrics we use sequential numbers: `rc/hawkular-cassandra-1`, which is the
master, `rc/hawkular-cassandra-2`, and so on.  I don't know enough about
clustering in elasticsearch (or even cassandra), but that's what the deployer
does currently.

bbguimaraes committed:
Improve default values documentation.

bbguimaraes commented:
@jcantrill: I believe I addressed all comments.  These playbooks are working
for a green-field installation, so it would be a good point to get it merged
before it gets (more) gigantic.  What do you think?
